File: packages/chp-sdk/src/merkle.ts
/**
 * RFC 6962 (Certificate Transparency) Merkle tree — the `chp-store-head-v2`
 * construction (proposal 0019). Domain-separated: leaf `SHA256(0x00‖data)`,
 * node `SHA256(0x01‖left‖right)`, odd sizes split at the largest power of two
 * `< n`. Byte-for-byte identical to Python `chp_core.merkle`.
 */

import { createHash } from 'node:crypto';

export const CHP_STORE_HEAD_V1 = 'chp-store-head-v1';
export const CHP_STORE_HEAD_V2 = 'chp-store-head-v2';

const _leafHash = (data: Buffer): Buffer =>
  createHash('sha256').update(Buffer.concat([Buffer.from([0]), data])).digest();
const _nodeHash = (l: Buffer, r: Buffer): Buffer =>
  createHash('sha256').update(Buffer.concat([Buffer.from([1]), l, r])).digest();

function _split(n: number): number {
  let k = 1;
  while (k * 2 < n) k *= 2;
  return k;
}

export function merkleRoot(leaves: Buffer[]): Buffer {
  const n = leaves.length;
  if (n === 0) return createHash('sha256').update(Buffer.alloc(0)).digest();
  if (n === 1) return _leafHash(leaves[0]);
  const k = _split(n);
  return _nodeHash(merkleRoot(leaves.slice(0, k)), merkleRoot(leaves.slice(k)));
}

export function inclusionProof(leaves: Buffer[], index: number): Buffer[] {
  const n = leaves.length;
  if (index < 0 || index >= n) throw new RangeError(`leaf index ${index} out of range for ${n}`);
  if (n === 1) return [];
  const k = _split(n);
  if (index < k) return [...inclusionProof(leaves.slice(0, k), index), merkleRoot(leaves.slice(k))];
  return [...inclusionProof(leaves.slice(k), index - k), merkleRoot(leaves.slice(0, k))];
}

// Recompute by replaying the split: the recursion consumes deeper path entries
// FIRST, then the sibling is shifted — bind the recursion result explicitly (do
// NOT rely on argument evaluation order).
function _walk(size: number, index: number, path: Buffer[], leaf: Buffer): Buffer {
  if (size === 1) return leaf;
  const k = _split(size);
  if (index < k) {
    const left = _walk(k, index, path, leaf);
    return _nodeHash(left, path.shift() as Buffer);
  }
  const right = _walk(size - k, index - k, path, leaf);
  return _nodeHash(path.shift() as Buffer, right);
}

export function verifyInclusion(
  root: Buffer, leafData: Buffer, index: number, treeSize: number, auditPath: Buffer[],
): boolean {
  if (index < 0 || index >= treeSize) return false;
  const path = [...auditPath];
  const computed = _walk(treeSize, index, path, _leafHash(leafData));
  return path.length === 0 && computed.equals(root);
}

// ── Consistency proofs (RFC 6962 §2.1.2, proposal 0022) ─────────────────────
// Prove a later tree (size n) is an append-only extension of an earlier one
// (size m ≤ n). Replay the prover's split so verify is the inverse of the build.

export function consistencyProof(leaves: Buffer[], m: number): Buffer[] {
  const n = leaves.length;
  if (m < 0 || m > n) throw new RangeError(`consistency: need 0 <= m(${m}) <= n(${n})`);
  if (m === 0 || m === n) return [];
  return _subproof(m, leaves, true);
}

function _subproof(m: number, leaves: Buffer[], b: boolean): Buffer[] {
  const n = leaves.length;
  if (m === n) return b ? [] : [merkleRoot(leaves)];
  const k = _split(n);
  if (m <= k) return [..._subproof(m, leaves.slice(0, k), b), merkleRoot(leaves.slice(k))];
  return [..._subproof(m - k, leaves.slice(k), false), merkleRoot(leaves.slice(0, k))];
}

// Returns [oldRoot@m, newRoot@n]; bind each shifted entry AFTER the recursion.
function _consistencyWalk(
  m: number, n: number, b: boolean, path: Buffer[], firstRoot: Buffer,
): [Buffer, Buffer] {
  if (m === n) {
    if (b) return [firstRoot, firstRoot];
    const h = path.shift() as Buffer;
    return [h, h];
  }
  const k = _split(n);
  if (m <= k) {
    const [old, newLeft] = _consistencyWalk(m, k, b, path, firstRoot);
    const right = path.shift() as Buffer;
    return [old, _nodeHash(newLeft, right)];
  }
  const [oldRight, newRight] = _consistencyWalk(m - k, n - k, false, path, firstRoot);
  const left = path.shift() as Buffer;
  return [_nodeHash(left, oldRight), _nodeHash(left, newRight)];
}

export function verifyConsistency(
  firstRoot: Buffer, secondRoot: Buffer, m: number, n: number, proof: Buffer[],
): boolean {
  if (m < 0 || m > n) return false;
  if (m === 0) return proof.length === 0;
  if (m === n) return proof.length === 0 && firstRoot.equals(secondRoot);
  const path = [...proof];
  const [old, next] = _consistencyWalk(m, n, true, path, firstRoot);
  return path.length === 0 && old.equals(firstRoot) && next.equals(secondRoot);
}

// ── Store-head schemes (chp-v0.2.md §12) ────────────────────────────────────

export type Leaves = Record<string, string | null>;

export function storeHeadLeaf(correlationId: string, headHash: string | null): Buffer {
  return Buffer.from(`${correlationId}\x00${headHash ?? ''}\n`, 'utf8');
}

export function storeHeadRoot(scheme: string, leaves: Leaves): string {
  const ordered = Object.keys(leaves).sort();
  if (scheme === CHP_STORE_HEAD_V1) {
    const h = createHash('sha256');
    for (const cid of ordered) h.update(storeHeadLeaf(cid, leaves[cid]));
    return h.digest('hex');
  }
  if (scheme === CHP_STORE_HEAD_V2) {
    return merkleRoot(ordered.map((cid) => storeHeadLeaf(cid, leaves[cid]))).toString('hex');
  }
  throw new Error(`unknown store-head scheme: ${scheme}`);
}

export function storeHeadSchemeMatching(leaves: Leaves, signedRoot: string): string | null {
  for (const scheme of [CHP_STORE_HEAD_V1, CHP_STORE_HEAD_V2]) {
    if (storeHeadRoot(scheme, leaves) === signedRoot) return scheme;
  }
  return null;
}

export interface StoreHeadInclusion {
  scheme: string;
  correlation_id: string;
  head_hash: string | null;
  leaf_index: number;
  tree_size: number;
  audit_path: string[];
}

export function storeHeadInclusionProof(leaves: Leaves, correlationId: string): StoreHeadInclusion {
  const ordered = Object.keys(leaves).sort();
  const index = ordered.indexOf(correlationId);
  if (index < 0) throw new Error(`correlation ${correlationId} not in leaves`);
  const path = inclusionProof(ordered.map((cid) => storeHeadLeaf(cid, leaves[cid])), index);
  return {
    scheme: CHP_STORE_HEAD_V2,
    correlation_id: correlationId,
    head_hash: leaves[correlationId],
    leaf_index: index,
    tree_size: ordered.length,
    audit_path: path.map((p) => p.toString('hex')),
  };
}

export function verifyStoreHeadInclusion(
  root: string, correlationId: string, headHash: string | null, proof: StoreHeadInclusion,
): boolean {
  if (proof.scheme !== CHP_STORE_HEAD_V2) return false;
  if (proof.correlation_id !== correlationId || proof.head_hash !== headHash) return false;
  try {
    const path = (proof.audit_path ?? []).map((h) => Buffer.from(h, 'hex'));
    return verifyInclusion(
      Buffer.from(root, 'hex'), storeHeadLeaf(correlationId, headHash),
      proof.leaf_index, proof.tree_size, path);
  } catch {
    return false;
  }
}

export interface StoreHeadConsistency {
  scheme: string;
  first_size: number;
  second_size: number;
  first_root: string;
  second_root: string;
  proof: string[];
}

export function storeHeadConsistencyProof(oldLeaves: Leaves, newLeaves: Leaves): StoreHeadConsistency {
  const oldOrdered = Object.keys(oldLeaves).sort();
  const newOrdered = Object.keys(newLeaves).sort();
  const newBytes = newOrdered.map((cid) => storeHeadLeaf(cid, newLeaves[cid]));
  return {
    scheme: CHP_STORE_HEAD_V2,
    first_size: oldOrdered.length,
    second_size: newOrdered.length,
    first_root: storeHeadRoot(CHP_STORE_HEAD_V2, oldLeaves),
    second_root: merkleRoot(newBytes).toString('hex'),
    proof: consistencyProof(newBytes, oldOrdered.length).map((h) => h.toString('hex')),
  };
}

export function verifyStoreHeadConsistency(
  oldRoot: string, newRoot: string, proof: StoreHeadConsistency,
): boolean {
  if (proof.scheme !== CHP_STORE_HEAD_V2) return false;
  if (proof.first_root !== oldRoot || proof.second_root !== newRoot) return false;
  try {
    const path = (proof.proof ?? []).map((h) => Buffer.from(h, 'hex'));
    return verifyConsistency(
      Buffer.from(oldRoot, 'hex'), Buffer.from(newRoot, 'hex'),
      proof.first_size, proof.second_size, path);
  } catch {
    return false;
  }
}

File: packages/python/chp_core/merkle.py
"""RFC 6962 (Certificate Transparency) Merkle tree — the `chp-store-head-v2`
construction (proposal 0019).

Domain-separated and second-preimage-safe: leaf hash ``SHA256(0x00 ‖ data)``,
interior node ``SHA256(0x01 ‖ left ‖ right)``, odd sizes split at the largest
power of two ``< n`` (RFC 6962 §2.1). It is the audited standard, so a Python, a
TypeScript, and a stdlib implementation compute the identical root and inclusion
proofs — the JCS/RFC-8785 discipline applied to trees.

Functions operate on ``bytes`` leaves; the store-head layer hex-encodes.
``verify_inclusion`` recomputes the root by replaying the SAME recursive split
the proof was built from — provably consistent with ``merkle_root`` /
``inclusion_proof`` rather than the error-prone fn/sn bit walk.
"""

from __future__ import annotations

import hashlib
from typing import Iterator


def _leaf_hash(data: bytes) -> bytes:
    return hashlib.sha256(b"\x00" + data).digest()


def _node_hash(left: bytes, right: bytes) -> bytes:
    return hashlib.sha256(b"\x01" + left + right).digest()


def _split(n: int) -> int:
    """The largest power of two strictly less than n (RFC 6962 §2.1; n >= 2)."""
    k = 1
    while k * 2 < n:
        k *= 2
    return k


def merkle_root(leaves: list[bytes]) -> bytes:
    """RFC 6962 Merkle Tree Hash over the ordered leaf byte-strings."""
    n = len(leaves)
    if n == 0:
        return hashlib.sha256(b"").digest()
    if n == 1:
        return _leaf_hash(leaves[0])
    k = _split(n)
    return _node_hash(merkle_root(leaves[:k]), merkle_root(leaves[k:]))


def inclusion_proof(leaves: list[bytes], index: int) -> list[bytes]:
    """RFC 6962 §2.1.1 audit path for leaf ``index`` — sibling subtree roots,
    ordered bottom-up (the leaf's immediate sibling first, the top sibling last)."""
    n = len(leaves)
    if not 0 <= index < n:
        raise IndexError(f"leaf index {index} out of range for {n} leaves")
    if n == 1:
        return []
    k = _split(n)
    if index < k:
        return inclusion_proof(leaves[:k], index) + [merkle_root(leaves[k:])]
    return inclusion_proof(leaves[k:], index - k) + [merkle_root(leaves[:k])]


def _walk(size: int, index: int, path: Iterator[bytes], leaf: bytes) -> bytes:
    if size == 1:
        return leaf
    k = _split(size)
    if index < k:
        left = _walk(k, index, path, leaf)
        return _node_hash(left, next(path))          # sibling = right subtree
    right = _walk(size - k, index - k, path, leaf)
    return _node_hash(next(path), right)             # sibling = left subtree


def verify_inclusion(root: bytes, leaf_data: bytes, index: int,
                     tree_size: int, audit_path: list[bytes]) -> bool:
    """True iff ``leaf_data`` at ``index`` is committed under ``root`` given
    ``audit_path`` (RFC 6962). Recomputes by replaying the build's split, so it
    is exactly the inverse of ``inclusion_proof``. All path entries must be
    consumed (a too-long or too-short path fails)."""
    if not 0 <= index < tree_size:
        return False
    it: Iterator[bytes] = iter(audit_path)
    try:
        computed = _walk(tree_size, index, it, _leaf_hash(leaf_data))
    except StopIteration:
        return False
    if next(it, None) is not None:
        return False  # leftover path entries
    return computed == root


# ── Store-head schemes (chp-v0.2.md §12) ────────────────────────────────────
# v1 = the flat SHA-256 fold; v2 = the RFC 6962 Merkle root (proposal 0019). One
# dispatcher so every recompute site agrees; raises on an unknown scheme (the §2
# canonicalization-dispatch pattern). The per-leaf bytes are identical in both.

CHP_STORE_HEAD_V1 = "chp-store-head-v1"
CHP_STORE_HEAD_V2 = "chp-store-head-v2"


def store_head_leaf(correlation_id: str, head_hash: str | None) -> bytes:
    """The per-correlation leaf bytes (identical under v1 and v2)."""
    return f"{correlation_id}\x00{head_hash or ''}\n".encode()


def store_head_root(scheme: str, leaves: dict) -> str:
    """The store-head root over ``{correlation_id: head_hash}``, hex. v1 folds;
    v2 builds the RFC 6962 Merkle root. Raises ``ValueError`` on an unknown
    scheme — a recompute under the wrong scheme cannot equal a signed root."""
    ordered = sorted(leaves)
    if scheme == CHP_STORE_HEAD_V1:
        h = hashlib.sha256()
        for cid in ordered:
            h.update(store_head_leaf(cid, leaves[cid]))
        return h.hexdigest()
    if scheme == CHP_STORE_HEAD_V2:
        return merkle_root([store_head_leaf(cid, leaves[cid]) for cid in ordered]).hex()
    raise ValueError(f"unknown store-head scheme: {scheme!r}")


def store_head_scheme_matching(leaves: dict, signed_root: str) -> str | None:
    """The store-head scheme whose recompute of ``leaves`` equals ``signed_root``,
    or None (tamper). Self-validating: because the signed root selects the scheme,
    a witness receipt need not carry the scheme — the witness header stays
    byte-identical for v1, and the audit tries each known scheme."""
    for scheme in (CHP_STORE_HEAD_V1, CHP_STORE_HEAD_V2):
        if store_head_root(scheme, leaves) == signed_root:
            return scheme
    return None


def store_head_inclusion_proof(leaves: dict, correlation_id: str) -> dict:
    """An RFC 6962 inclusion proof (chp-store-head-v2) that ``correlation_id``'s
    leaf is committed under the Merkle root — the `store-head-inclusion` object."""
    ordered = sorted(leaves)
    if correlation_id not in leaves:
        raise KeyError(correlation_id)
    index = ordered.index(correlation_id)
    leaf_bytes = [store_head_leaf(cid, leaves[cid]) for cid in ordered]
    path = inclusion_proof(leaf_bytes, index)
    return {
        "scheme": CHP_STORE_HEAD_V2,
        "correlation_id": correlation_id,
        "head_hash": leaves[correlation_id],
        "leaf_index": index,
        "tree_size": len(ordered),
        "audit_path": [p.hex() for p in path],
    }


def verify_store_head_inclusion(root: str, correlation_id: str,
                                head_hash: str | None, proof: dict) -> bool:
    """Third-party, witness-free (§12, proposal 0019): recompute the Merkle root
    from a single correlation's leaf up ``proof.audit_path`` and check it equals
    the signed/anchored ``root`` — with NO leaves snapshot. The proof binds the
    leaf bytes, so a forged head_hash or wrong correlation_id fails."""
    if proof.get("scheme") != CHP_STORE_HEAD_V2:
        return False
    if proof.get("correlation_id") != correlation_id or proof.get("head_hash") != head_hash:
        return False
    try:
        path = [bytes.fromhex(h) for h in proof.get("audit_path", [])]
        return verify_inclusion(
            bytes.fromhex(root), store_head_leaf(correlation_id, head_hash),
            int(proof["leaf_index"]), int(proof["tree_size"]), path)
    except (ValueError, KeyError, TypeError):
        return False


def _selfcheck() -> None:
    """Assertion-based sanity check (RFC 6962 known-answers + inclusion round
    trips). Runnable: ``python -m chp_core.merkle``."""
    # Empty tree: SHA256 of the empty string (the RFC 6962 fixed point).
    assert merkle_root([]).hex() == \
        "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
    # Single leaf: SHA256(0x00 ‖ data).
    assert merkle_root([b"L0"]) == _leaf_hash(b"L0")
    # Two leaves: SHA256(0x01 ‖ leaf(L0) ‖ leaf(L1)).
    assert merkle_root([b"L0", b"L1"]) == _node_hash(_leaf_hash(b"L0"), _leaf_hash(b"L1"))
    # Odd (3) leaves split 2|1 (largest power of two < 3 is 2).
    three = [b"a", b"b", b"c"]
    assert merkle_root(three) == _node_hash(
        _node_hash(_leaf_hash(b"a"), _leaf_hash(b"b")), _leaf_hash(b"c"))
    # Every leaf's inclusion proof round-trips for sizes 1..9; a forged leaf fails.
    for n in range(1, 10):
        leaves = [f"leaf-{i}".encode() for i in range(n)]
        root = merkle_root(leaves)
        for i in range(n):
            proof = inclusion_proof(leaves, i)
            assert verify_inclusion(root, leaves[i], i, n, proof), (n, i)
            assert not verify_inclusion(root, b"FORGED", i, n, proof), (n, i)
            if proof:  # a tampered path fails
                bad = list(proof); bad[0] = b"\x00" * 32
                assert not verify_inclusion(root, leaves[i], i, n, bad), (n, i)
    print("merkle self-check OK (RFC 6962)")


if __name__ == "__main__":
    _selfcheck()
